### PR TITLE
[MODORDERS-800] - Implement moving encumbered value from budget 1 to budget 2 (FLOW #2)

### PR DESCRIPTION
**Pull request by Abdulkhakimov**
## Approach

- Catch invoice related errors after encumbrances will be processed
- Fix NPE bugs

## Learning
[MODORDERS-800](https://issues.folio.org/browse/MODORDERS-800)

## Pre-Merge Checklist:
Before merging this PR, please go through the following list and take appropriate actions.

- Does this PR meet or exceed the expected quality standards?
  - [ ] Code coverage on new code is 80% or greater
  - [x] Duplications on new code is 3% or less
  - [x] There are no major code smells or security issues
- Does this introduce breaking changes?
  - [ ] Were any API paths or methods changed, added or removed?
  - [ ] Were there any schema changes?
  - [ ] Did any of the interface versions change?
  - [ ] Were permissions changed, added, or removed?
  - [ ] Are there new interface dependencies?
  - [x] There are no breaking changes in this PR.
  
If there are breaking changes, please **STOP** and consider the following:

- What other modules will these changes impact?
- Do JIRAs exist to update the impacted modules?
  - [ ] If not, please create them
  - [ ] Do they contain the appropriate level of detail?  Which endpoints/schemas changed, etc.
  - [ ] Do they have all they appropriate links to blocked/related issues?
- Are the JIRAs under active development?  
  - [ ] If not, contact the project's PO and make sure they're aware of the urgency.
- Do PRs exist for these changes?
  - [ ] If so, have they been approved?

Ideally all of the PRs involved in breaking changes would be merged in the same day to avoid breaking the folio-testing environment.  Communication is paramount if that is to be achieved, especially as the number of intermodule and inter-team dependencies increase.  

While it's helpful for reviewers to help identify potential problems, ensuring that it's safe to merge is ultimately the responsibility of the PR assignee.

### CODE REVIEW

File: src/main/java/org/folio/helper/PurchaseOrderLineHelper.java [hunk 1]
```diff
@@ -808,11 +808,18 @@ private Future<Void> processPoLineEncumbrances(CompositePurchaseOrder compOrder,
         OrderWorkflowType.PENDING_TO_PENDING : OrderWorkflowType.PENDING_TO_OPEN;
       EncumbranceWorkflowStrategy strategy = encumbranceWorkflowStrategyFactory.getStrategy(workflowType);
       CompositePurchaseOrder poFromStorage = JsonObject.mapFrom(compOrder).mapTo(CompositePurchaseOrder.class);
-      return strategy.processEncumbrances(compOrder.withCompositePoLines(Collections.singletonList(compositePoLine)), poFromStorage, requestContext);
+      return strategy.processEncumbrances(compOrder.withCompositePoLines(Collections.singletonList(compositePoLine)), poFromStorage, requestContext)
+        .recover(throwable -> isInvoiceRelatedError(workflowType, throwable) ? Future.succeededFuture() : Future.failedFuture(throwable));
```
Comment (SerhiiNosko) [inline src/main/java/org/folio/helper/PurchaseOrderLineHelper.java:812]: We should not rely our business logic on handling error as in this code, if there is some invoice related error - we should not assume that edit was succesful, provided approach should be improved

Comment (Abdulkhakimov) [reply src/main/java/org/folio/helper/PurchaseOrderLineHelper.java]: Done

File: src/main/java/org/folio/models/PoLineInvoiceLineHolder.java [hunk 2]
```diff
@@ -5,23 +5,24 @@
...
 
 public class PoLineInvoiceLineHolder {
-  private CompositePoLine poLine;
-  private JsonObject poLineFromStorage;
+  private CompositePoLine poLineFromRequest;
+  private PoLine poLineFromStorage;
   private List<InvoiceLine> invoiceLines;
   private List<InvoiceLine> openOrReviewedInvoiceLines;
-  private List<InvoiceLine> currentYearPaidInvoiceLines;
+  private List<InvoiceLine> currentYearPaidOrCancelledInvoiceLines;
```
Comment (SerhiiNosko) [inline src/main/java/org/folio/models/PoLineInvoiceLineHolder.java:18]: I think to simplify naming we can use paidOrCancelledInvoiceLines name

Comment (Abdulkhakimov) [reply src/main/java/org/folio/models/PoLineInvoiceLineHolder.java]: Done

File: src/main/java/org/folio/service/invoice/POLInvoiceLineRelationService.java [hunk 3]
```diff
@@ -1,34 +1,60 @@
...
+      .map(Transaction::getEncumbrance).map(Encumbrance::getSourcePoLineId).distinct().collect(Collectors.toList());
+    return invoiceLineService.getInvoiceLinesByOrderLineIds(poLineIds, requestContext)
+      .map(PoLineInvoiceLineHolderBuilder::getInvoiceLinesWithTransactions)
+      .map(invoiceLines -> invoiceLines.stream().flatMap(invoiceLine -> invoiceLine.getFundDistributions().stream())
+        .map(org.folio.rest.acq.model.invoice.FundDistribution::getEncumbrance).collect(Collectors.toSet()))
+      .map(encumbranceIds -> holders.stream().filter(holder -> !encumbranceIds.contains(holder.getOldEncumbrance().getId())).collect(Collectors.toList()))
+      .map(encumbrancesProcessingHolder::withEncumbrancesForDelete);
+  }
+
+  private List<String> getFundIdFromFundDistribution(List<FundDistribution> fundDistributions) {
```
[the remark this answers is not in the document]
Comment (Abdulkhakimov) [reply src/main/java/org/folio/service/invoice/POLInvoiceLineRelationService.java]: Done

**Review by Dmitriy-Butramyou — APPROVED** [review]

**Review by SerhiiNosko — APPROVED** [review]

**Pull request merged**